injection/sharedmain/main.go

```diff
@@ -143,102 +166,180 @@ func MainWithConfig(ctx context.Context, component string, cfg *rest.Config, cto
+		<-egCtx.Done()
 
-	eg, egCtx := errgroup.WithContext(ctx)
-	eg.Go(profilingServer.ListenAndServe)
+		profilingServer.Shutdown(context.Background())
+		// Don't forward ErrServerClosed as that indicates we're already shutting down.
+		if err := eg.Wait(); err != nil && err != http.ErrServerClosed {
+			logger.Errorw("Error while running server", zap.Error(err))
+		}
+	}
 
-	// If we have one or more admission controllers, then start the webhook
-	// and pass them in.
-	if len(webhooks) > 0 {
-		// Register webhook metrics
-		webhook.RegisterMetrics()
+	recorder := controller.GetEventRecorder(ctx)
+	if recorder == nil {
+		// Create event broadcaster
+		logger.Debug("Creating event broadcaster")
+		eventBroadcaster := record.NewBroadcaster()
+		watches := []watch.Interface{
+			eventBroadcaster.StartLogging(logger.Named("event-broadcaster").Infof),
+			eventBroadcaster.StartRecordingToSink(
+				&typedcorev1.EventSinkImpl{Interface: kubeclient.Get(ctx).CoreV1().Events("")}),
+		}
+		recorder = eventBroadcaster.NewRecorder(
+			// todo: what agent name?
+			scheme.Scheme, corev1.EventSource{Component: "fake-agent-name"})
+		go func() {
+			<-ctx.Done()
+			for _, w := range watches {
+				w.Stop()
+			}
+		}()
+	}
 
-		wh, err := webhook.New(ctx, webhooks)
+	// Set up leader election config
+	leaderElectionConfig, err := GetLeaderElectionConfig(ctx)
+	if err != nil {
+		log.Fatalf("Error loading leader election configuration: %v", err)
+	}
+	leConfig := leaderElectionConfig.GetComponentConfig(component)
+
+	if !leConfig.LeaderElect {
+		log.Printf("%v will not run in leader-elected mode", component)
+		run(ctx)
+		panic("unreachable")
+	} else {
+		// create a unique identifier so that two controllers on the same host don't
+		// race.
+		id, err := os.Hostname()
+		if err != nil {
+			logger.Fatalw("Failed to get hostname for leader election", zap.Error(err))
+		}
+		id = id + "_" + string(uuid.NewUUID())
+		log.Printf("%v will run in leader-elected mode with id %v", component, id)
+
+		rl, err := resourcelock.New(leConfig.ResourceLock,
+			system.Namespace(), // use namespace we are running in
+			component,          // component is used as the resource name
+			kubeclient.Get(ctx).CoreV1(),
+			kubeclient.Get(ctx).CoordinationV1(),
+			resourcelock.ResourceLockConfig{
+				Identity:      id,
+				EventRecorder: recorder,
+			})
 		if err != nil {
-			logger.Fatalw("Failed to create webhook", zap.Error(err))
+			logger.Fatalw("error creating lock: %v", err)
 		}
-		eg.Go(func() error {
-			return wh.Run(ctx.Done())
+
+		leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+			Lock:          rl,
+			LeaseDuration: leConfig.LeaseDuration,
+			RenewDeadline: leConfig.RenewDeadline,
+			RetryPeriod:   leConfig.RetryPeriod,
+			Callbacks: leaderelection.LeaderCallbacks{
+				OnStartedLeading: run,
+				OnStoppedLeading: func() {
+					logger.Fatalw("leaderelection lost")
+				},
+			},
+			// TODO: investigate using watchdog
```

Comment:
Could we add an issue to track this and add a pointer to that here?

Reply:
Opened https://github.com/knative/pkg/issues/1048

Reply:
+1